rust/src/utils.rs

```diff
@@ -304,6 +310,101 @@ pub(crate) fn translate_path_for_ostree_impl(path: &str) -> Option<String> {
+    if rpmostree_tmpfiles_list.is_empty() || system_tmpfiles_list.is_empty() {
+        println!("Not found any auto-generated or system tmpfiles.d config");
+        return Ok(());
+    }
+
+    // save rpm-ostree auto generated entries to hashmaps
+    let tmpfiles_dir = tmprootfs_dfd
+        .open_dir(RPMOSTREE_TMPFILESD)
+        .context("readdir {RPMOSTREE_TMPFILESD}")?;
+    let mut rpmostree_tmpfiles_entries =
+        save_tmpfile_entries(&tmpfiles_dir, &rpmostree_tmpfiles_list).unwrap();
+    // save system entries to hashmaps
+    let tmpfiles_dir = tmprootfs_dfd
+        .open_dir(TMPFILESD)
+        .context("readdir {TMPFILESD}")?;
+    let system_tmpfiles_entries =
+        save_tmpfile_entries(&tmpfiles_dir, &system_tmpfiles_list).unwrap();
+
+    // remove duplicated entries in auto-generated tmpfiles.d,
+    // which are already in system tmpfiles
+    for key in system_tmpfiles_entries.keys() {
+        if rpmostree_tmpfiles_entries.contains_key(key) {
+            rpmostree_tmpfiles_entries.remove(key);
+        }
+    }
+
+    {
+        // save the noduplicated entries
+        let mut entries = String::from("");
+        for (_key, value) in rpmostree_tmpfiles_entries {
+            entries.push_str(&format!("{}\n", value));
+        }
+
+        if tmpfiles_dir.try_exists(AUTOVAR_PATH)? {
+            tmpfiles_dir.remove_file(AUTOVAR_PATH)?
+        }
+        let perms = Permissions::from_mode(0o644);
+        tmpfiles_dir.atomic_write_with_perms(&AUTOVAR_PATH, entries.as_bytes(), perms)?;
+    }
+    Ok(())
+}
+
+#[context("Get tmpfiles files list")]
+pub(crate) fn get_tmpfiles_path_list(tmprootfs_dfd: &Dir, path: &str) -> Result<Vec<String>> {
+    let tmpfiles_dir = tmprootfs_dfd.open_dir(path).context("readdir {path}")?;
+    let mut tmpfiles_list = Vec::new();
+    for entry in tmpfiles_dir.entries()? {
+        let entry = entry?;
+        let name = entry.file_name();
+        let name = name.to_str().unwrap();
+        // skip README
+        if name == "README" || name == "rpm-ostree-autovar.conf" {
+            continue;
+        }
+        tmpfiles_list.push(format!("{name}"));
+    }
+    Ok(tmpfiles_list)
+}
+
+#[context("Save tmpfiles entries to hashmap")]
+pub(crate) fn save_tmpfile_entries(
+    tmpfiles_dir: &Dir,
+    var_list: &Vec<String>,
+) -> Result<HashMap<String, String>> {
+    let mut tmpflies_entries: HashMap<String, String> = HashMap::new();
+    for tmpflies in var_list.iter() {
+        let contents = tmpfiles_dir.open(tmpflies).map(BufReader::new)?;
+
+        for (line_num, line) in contents.lines().enumerate() {
+            let input = line
+                .with_context(|| format!("failed to read tmpfiles entry at line {}", line_num))?;
+
+            // Skip empty and comment lines
+            if input.is_empty() || input.starts_with('#') {
+                continue;
+            }
+
+            let parts: Vec<&str> = input.split_whitespace().collect();
```

Comment:
Hmm, when scanning the canonical dir, we don't actually need to store the other parts of the line. Maybe we should return a `HashMap<String, Option<String>>` instead and add an argument like `keep_line`?

Alternatively, a more Rust-y way to do this is to convert `save_tmpfile_entries()` to an iterator over the entries (e.g. `scan_tmpfile_entries()`); then in the canonical scan, we `.collect()` into a `HashSet` and in the rpm-ostree scan, a `HashMap`.

Reply:
Will try about it, thanks!